File: HMCL/src/main/java/org/jackhuang/hmcl/ui/main/SettingsPage.java
/*
 * Hello Minecraft! Launcher
 * Copyright (C) 2020  huangyuhui <huanghongxun2008@126.com> and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.jackhuang.hmcl.ui.main;

import com.jfoenix.controls.JFXButton;
import javafx.beans.InvalidationListener;
import javafx.beans.WeakInvalidationListener;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.StringProperty;
import javafx.css.PseudoClass;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.control.Label;
import javafx.scene.control.ScrollPane;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import org.jackhuang.hmcl.task.Schedulers;
import org.jackhuang.hmcl.ui.Controllers;
import org.jackhuang.hmcl.ui.FXUtils;
import org.jackhuang.hmcl.ui.SVG;
import org.jackhuang.hmcl.ui.construct.*;
import org.jackhuang.hmcl.ui.construct.MessageDialogPane.MessageType;
import org.jackhuang.hmcl.upgrade.RemoteVersion;
import org.jackhuang.hmcl.upgrade.UpdateChannel;
import org.jackhuang.hmcl.upgrade.UpdateChecker;
import org.jackhuang.hmcl.upgrade.UpdateHandler;
import org.jackhuang.hmcl.util.Lang;
import org.jackhuang.hmcl.util.LauncherLogExporter;
import org.jackhuang.hmcl.util.StringUtils;
import org.jackhuang.hmcl.util.i18n.I18n;
import org.jackhuang.hmcl.util.i18n.SupportedLocale;

import java.util.List;
import java.util.concurrent.CompletableFuture;

import static org.jackhuang.hmcl.setting.SettingsManager.settings;
import static org.jackhuang.hmcl.util.i18n.I18n.i18n;
import static org.jackhuang.hmcl.util.logging.Logger.LOG;

public final class SettingsPage extends ScrollPane {
    @SuppressWarnings("FieldCanBeLocal")
    private final InvalidationListener updateListener;

    public SettingsPage() {
        this.setFitToWidth(true);

        VBox rootPane = new VBox(10);
        rootPane.setPadding(new Insets(10));
        this.setContent(rootPane);
        FXUtils.smoothScrolling(this);

        {
            ComponentList updatePaneList = new ComponentList();
            {
                ObjectProperty<UpdateChannel> updateChannel;
                {

                    JFXButton updateButton = FXUtils.newToggleButton4(SVG.UPDATE, 20);
                    updateButton.setOnAction(e -> onUpdate());
                    updateButton.setPadding(Insets.EMPTY);
                    FXUtils.installFastTooltip(updateButton, i18n("update.tooltip"));

                    var updatePane = new LineSelectButton<UpdateChannel>() {

                        {
                            getStyleClass().add("update-pane");
                            setNode(IDX_TRAILING, updateButton);
                        }

                        @Override
                        protected int getTrailingTextIndex() {
                            return LineComponent.IDX_TRAILING + 1;
                        }
                    };
                    updateChannel = updatePane.valueProperty();
                    updatePane.setTitle(i18n("update"));
                    updatePane.setValue(UpdateChannel.getChannel());

                    updatePane.setNullSafeConverter(channel -> i18n("update.channel." + channel.channelName));
                    updatePane.setItems(List.of(UpdateChannel.STABLE, UpdateChannel.DEVELOPMENT));
                    updatePane.setDescriptionConverter(channel -> i18n("update.note." + channel.channelName));

                    final StringProperty lblUpdateSubProperty = updatePane.subtitleProperty();

                    {
                        updateListener = any -> {
                            boolean outdated = UpdateChecker.isOutdated();

                            updateButton.setVisible(outdated);
                            updateButton.setManaged(outdated);
                            updatePane.pseudoClassStateChanged(PseudoClass.getPseudoClass("active"), outdated);

                            if (UpdateChecker.isOutdated()) {
                                lblUpdateSubProperty.set(i18n("update.newest_version", UpdateChecker.getLatestVersion().version()));
                            } else if (UpdateChecker.isCheckingUpdate()) {
                                lblUpdateSubProperty.set(i18n("update.checking"));
                            } else {
                                lblUpdateSubProperty.set(i18n("update.latest"));
                            }
                        };
                        UpdateChecker.latestVersionProperty().addListener(new WeakInvalidationListener(updateListener));
                        UpdateChecker.outdatedProperty().addListener(new WeakInvalidationListener(updateListener));
                        UpdateChecker.checkingUpdateProperty().addListener(new WeakInvalidationListener(updateListener));
                        updateListener.invalidated(null);
                    }

                    updatePaneList.getContent().add(updatePane);
                }

                {
                    LineToggleButton previewPane = new LineToggleButton();
                    previewPane.setTitle(i18n("update.preview"));
                    previewPane.setSubtitle(i18n("update.preview.subtitle"));
                    previewPane.selectedProperty().bindBidirectional(settings().acceptPreviewUpdateProperty());

                    InvalidationListener checkUpdateListener = e -> {
                        UpdateChecker.requestCheckUpdate(updateChannel.get(), previewPane.isSelected());
                    };
                    updateChannel.addListener(checkUpdateListener);
                    previewPane.selectedProperty().addListener(checkUpdateListener);

                    updatePaneList.getContent().add(previewPane);
                }

                {
                    LineToggleButton disableAutoShowUpdateDialogPane = new LineToggleButton();
                    disableAutoShowUpdateDialogPane.setTitle(i18n("update.disable_auto_show_update_dialog"));
                    disableAutoShowUpdateDialogPane.setSubtitle(i18n("update.disable_auto_show_update_dialog.subtitle"));
                    disableAutoShowUpdateDialogPane.selectedProperty().bindBidirectional(settings().disableAutoShowUpdateDialogProperty());
                    updatePaneList.getContent().add(disableAutoShowUpdateDialogPane);
                }

                rootPane.getChildren().addAll(ComponentList.createComponentListTitle(i18n("update")), updatePaneList);
            }

            {
                ComponentList languagePaneList = new ComponentList();

                {
                    var chooseLanguagePane = new LineSelectButton<SupportedLocale>();
                    chooseLanguagePane.setTitle(i18n("settings.launcher.language"));
                    chooseLanguagePane.setSubtitle(i18n("settings.take_effect_after_restart"));

                    SupportedLocale currentLocale = I18n.getLocale();
                    chooseLanguagePane.setNullSafeConverter(locale -> {
                        if (locale.isDefault())
                            return locale.getDisplayName(currentLocale);
                        else if (locale.isSameLanguage(currentLocale))
                            return locale.getDisplayName(locale);
                        else
                            return locale.getDisplayName(currentLocale) + " - " + locale.getDisplayName(locale);
                    });
                    chooseLanguagePane.setItems(SupportedLocale.getSupportedLocales());
                    chooseLanguagePane.valueProperty().bindBidirectional(settings().languageProperty());

                    languagePaneList.getContent().add(chooseLanguagePane);

                }

                rootPane.getChildren().addAll(ComponentList.createComponentListTitle(i18n("settings.launcher.language")), languagePaneList);
            }

            {
                ComponentList miscPaneList = new ComponentList();

                {
                    LineToggleButton disableAprilFools = new LineToggleButton();
                    disableAprilFools.setTitle(i18n("settings.launcher.disable_april_fools"));
                    disableAprilFools.setSubtitle(i18n("settings.take_effect_after_restart"));
                    disableAprilFools.selectedProperty().bindBidirectional(settings().disableAprilFoolsProperty());
                    miscPaneList.getContent().add(disableAprilFools);
                }

                {
                    BorderPane debugPane = new BorderPane();

                    Label left = new Label(i18n("settings.launcher.debug"));
                    BorderPane.setAlignment(left, Pos.CENTER_LEFT);
                    debugPane.setLeft(left);

                    JFXButton openLogFolderButton = new JFXButton(i18n("settings.launcher.launcher_log.reveal"));
                    openLogFolderButton.setOnAction(e -> openLogFolder());
                    openLogFolderButton.getStyleClass().add("jfx-button-border");
                    if (LOG.getLogFile() == null)
                        openLogFolderButton.setDisable(true);

                    SpinnerPane exportLogPane = new SpinnerPane();

                    JFXButton logButton = FXUtils.newBorderButton(i18n("settings.launcher.launcher_log.export"));
                    exportLogPane.setContent(logButton);
                    logButton.setOnAction(e -> {
                        exportLogPane.showSpinner();
                        CompletableFuture.supplyAsync(Lang.wrap(LauncherLogExporter::exportLogsAsZip), Schedulers.io()).whenCompleteAsync((result, exception) -> {
                            exportLogPane.hideSpinner();
                            if (exception == null) {
                                Controllers.dialog(i18n("settings.launcher.launcher_log.export.success", result));
                                FXUtils.showFileInExplorer(result);
                            } else {
                                LOG.warning("Failed to export logs", exception);
                                Controllers.dialog(
                                        i18n("settings.launcher.launcher_log.export.failed") + "\n" + StringUtils.getStackTrace(exception),
                                        null,
                                        MessageType.ERROR
                                );
                            }
                        }, Schedulers.javafx());
                    });

                    HBox buttonBox = new HBox();
                    buttonBox.setSpacing(10);
                    buttonBox.getChildren().addAll(openLogFolderButton, exportLogPane);
                    BorderPane.setAlignment(buttonBox, Pos.CENTER_RIGHT);
                    debugPane.setRight(buttonBox);

                    miscPaneList.getContent().add(debugPane);
                }

                rootPane.getChildren().addAll(ComponentList.createComponentListTitle(i18n("settings.launcher.misc")), miscPaneList);
            }
        }
    }

    private void openLogFolder() {
        FXUtils.openFolder(LOG.getLogFile().getParent());
    }

    private void onUpdate() {
        RemoteVersion target = UpdateChecker.getLatestVersion();
        if (target == null) {
            return;
        }
        UpdateHandler.updateFrom(target);
    }
}

File: HMCL/src/main/kotlin/org/jackhuang/hmcl/ui/SettingsPage.kt
/*
 * Hello Minecraft! Launcher.
 * Copyright (C) 2017  huangyuhui <huanghongxun2008@126.com>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see {http://www.gnu.org/licenses/}.
 */
package org.jackhuang.hmcl.ui

import com.jfoenix.controls.JFXComboBox
import com.jfoenix.controls.JFXTextField
import javafx.beans.property.SimpleStringProperty
import javafx.beans.property.StringProperty
import javafx.collections.FXCollections
import javafx.fxml.FXML
import javafx.scene.control.Label
import javafx.scene.layout.StackPane
import javafx.scene.text.Font
import org.jackhuang.hmcl.i18n
import org.jackhuang.hmcl.setting.DownloadProviders
import org.jackhuang.hmcl.setting.Locales
import org.jackhuang.hmcl.setting.Proxies
import org.jackhuang.hmcl.setting.Settings
import org.jackhuang.hmcl.ui.construct.FileItem
import org.jackhuang.hmcl.ui.construct.FontComboBox
import org.jackhuang.hmcl.ui.construct.Validator
import org.jackhuang.hmcl.ui.wizard.DecoratorPage
import org.jackhuang.hmcl.util.onChange

class SettingsPage : StackPane(), DecoratorPage {
    private val titleProperty: StringProperty = SimpleStringProperty(this, "title", i18n("launcher.title.launcher"))

    override fun titleProperty() = titleProperty

    @FXML lateinit var txtProxyHost: JFXTextField
    @FXML lateinit var txtProxyPort: JFXTextField
    @FXML lateinit var txtProxyUsername: JFXTextField
    @FXML lateinit var txtProxyPassword: JFXTextField
    @FXML lateinit var cboProxyType: JFXComboBox<*>
    @FXML lateinit var cboFont: FontComboBox
    @FXML lateinit var cboLanguage: JFXComboBox<*>
    @FXML lateinit var cboDownloadSource: JFXComboBox<*>
    @FXML lateinit var fileCommonLocation: FileItem
    @FXML lateinit var fileBackgroundLocation: FileItem
    @FXML lateinit var lblDisplay: Label
    @FXML lateinit var txtFontSize: JFXTextField

    init {
        loadFXML("/assets/fxml/setting.fxml")

        cboLanguage.limitWidth(400.0)
        cboDownloadSource.limitWidth(400.0)

        txtProxyHost.text = Settings.INSTANCE.proxyHost
        txtProxyHost.textProperty().onChange { Settings.INSTANCE.proxyHost = it }

        txtProxyPort.text = Settings.INSTANCE.proxyPort
        txtProxyPort.textProperty().onChange { Settings.INSTANCE.proxyPort = it }

        txtProxyUsername.text = Settings.INSTANCE.proxyUser
        txtProxyUsername.textProperty().onChange { Settings.INSTANCE.proxyUser = it }

        txtProxyPassword.text = Settings.INSTANCE.proxyPass
        txtProxyPassword.textProperty().onChange { Settings.INSTANCE.proxyPass = it }

        cboDownloadSource.selectionModel.select(DownloadProviders.DOWNLOAD_PROVIDERS.indexOf(Settings.INSTANCE.downloadProvider))
        cboDownloadSource.selectionModel.selectedIndexProperty().onChange {
            Settings.INSTANCE.downloadProvider = DownloadProviders.getDownloadProvider(it)
        }

        cboFont.selectionModel.select(Settings.INSTANCE.font.family)
        cboFont.valueProperty().onChange {
            val font = Font.font(it, Settings.INSTANCE.font.size)
            Settings.INSTANCE.font = font
            lblDisplay.style = "-fx-font: ${Settings.INSTANCE.font.size} \"${font.family}\";"
        }
        txtFontSize.text = Settings.INSTANCE.font.size.toString()
        txtFontSize.validators += Validator { it.toDoubleOrNull() != null }
        txtFontSize.textProperty().onChange {
            if (txtFontSize.validate()) {
                val font = Font.font(Settings.INSTANCE.font.family, it!!.toDouble())
                Settings.INSTANCE.font = font
                lblDisplay.style = "-fx-font: ${font.size} \"${Settings.INSTANCE.font.family}\";"
            }
        }
        lblDisplay.style = "-fx-font: ${Settings.INSTANCE.font.size} \"${Settings.INSTANCE.font.family}\";"

        val list = FXCollections.observableArrayList<Label>()
        for (locale in Locales.LOCALES) {
            list += Label(locale.getName(Settings.INSTANCE.locale.resourceBundle))
        }
        cboLanguage.items = list
        cboLanguage.selectionModel.select(Locales.LOCALES.indexOf(Settings.INSTANCE.locale))
        cboLanguage.selectionModel.selectedIndexProperty().onChange {
            Settings.INSTANCE.locale = Locales.getLocale(it)
        }

        cboProxyType.selectionModel.select(Proxies.PROXIES.indexOf(Settings.INSTANCE.proxyType))
        cboProxyType.selectionModel.selectedIndexProperty().onChange {
            Settings.INSTANCE.proxyType = Proxies.getProxyType(it)
        }

        fileCommonLocation.setProperty(Settings.INSTANCE.commonPathProperty())
    }


}